centaur/src/main/scala/centaur/test/Test.scala

```diff
@@ -419,6 +425,104 @@ object Operations extends StrictLogging {
+    def replaceVariables(value: JsValue): JsValue = value match {
+      case s: JsString => JsString(s.value.replaceAll("<<UUID>>", workflowId.toString).replaceAll("<<WORKFLOW_ROOT>>", workflowRoot).replaceFirst("^~>", ""))
+      case o => o
+    }
+    val filterLabels: PartialFunction[(String, JsValue), (String, JsValue)] = {
+      case (k, v) if k.startsWith(prefix) => k.substring(prefix.length) -> replaceVariables(v)
+    }
+
+    for {
+      flat <- workflow.metadata.toList
+      selected <- flat.value.toList collect filterLabels
+    } yield selected
+  }
+
+  def validateOutputs(submittedWorkflow: SubmittedWorkflow,
+                      workflow: Workflow,
+                      workflowRoot: String): Test[Unit] = new Test[Unit] {
+
+    override def run: IO[Unit] = {
+      import centaur.test.metadata.WorkflowFlatOutputs._
+
+      val expectedOutputs: List[(String, JsValue)] = selectAndProcessExpectationFlatMetadata(workflow, "outputs.", submittedWorkflow.id, workflowRoot)
+      val ioActualOutputs: IO[Map[String, JsValue]] = CentaurCromwellClient.outputs(submittedWorkflow) map { _.asFlat.stringifyValues }
+
+      ioActualOutputs flatMap { actualOutputs =>
+        val expected = expectedOutputs.toSet
+        val actual = actualOutputs.toSet
+
+        lazy val inActualButNotInExpected = actual.diff(expected)
+        lazy val inExpectedButNotInActual = expected.diff(actual)
+
+        if (!workflow.allowOtherOutputs && inActualButNotInExpected.nonEmpty) {
+          val message = s"In actual outputs but not in expected and other outputs not allowed: ${inActualButNotInExpected.mkString(", ")}"
+          IO.raiseError(CentaurTestException(message, workflow, submittedWorkflow))
+        } else if (inExpectedButNotInActual.nonEmpty) {
+          val message = s"In expected outputs but not in actual: ${inExpectedButNotInActual.mkString(", ")}"
+          IO.raiseError(CentaurTestException(message, workflow, submittedWorkflow))
+        } else {
+          IO.unit
+        }
+      }
+    }
+  }
+
+  def validateLabels(submittedWorkflow: SubmittedWorkflow,
+                     workflow: Workflow,
+                     workflowRoot: String): Test[Unit] = new Test[Unit] {
+    override def run: IO[Unit] = {
+      import centaur.test.metadata.WorkflowFlatLabels._
+
+      val workflowIdLabel = ("cromwell-workflow-id", JsString(s"cromwell-${submittedWorkflow.id}"))
+      val expectedLabels: List[(String, JsValue)] = workflowIdLabel ::
+        selectAndProcessExpectationFlatMetadata(workflow, "labels.", submittedWorkflow.id, workflowRoot)
+      val ioActualLabels: IO[Map[String, JsValue]] = CentaurCromwellClient.labels(submittedWorkflow) map { _.asFlat.stringifyValues }
+
+      ioActualLabels flatMap { actualLabels =>
+        val diff = expectedLabels.toSet.diff(actualLabels.toSet)
```

Comment:
Is it possible that there are more labels than expected (like with outputs)?

Reply:
Probably but I'm not sure. I am more sure that there isn't a label equivalent of `allowOtherOutputs` though.